oio/common/cryptography_tools.py

```diff
@@ -0,0 +1,190 @@
+                if fd_out:
+                    fd_out.write(to_write)
+                    fd_out.flush()
+                else:
+                    content += to_write
+                yield self._generate_dict(False, bytes_transferred,
+                                          max_bytes_read, content)
+                f(remains)
+                if fd_out:
+                    fd_out.write(remains)
+                else:
+                    content = remains
+                file_transferred = len(remains)
+                nb_file += 1
+                # else, we just update the data
+            else:
+                f(cipher)
+                if fd_out:
+                    fd_out.write(cipher)
+                else:
+                    content += cipher
+                file_transferred += len(cipher)
+                bytes_transferred += len(data)
+            if fd_out:
+                fd_out.flush()
+            # False is just the end of the file
+            yield self._generate_dict(True, len(data),
+                                      file_transferred, content)
+
+    # Takes a fd and decrypt the content of ciphered content in the descriptor
+    # yielding every tokens available
+    # one hook available :
+    # buffer_to_tokens(stream) : generator
+    # -> isolate the fd into multiple tokens to be decrypted
+    def decrypt_from_buffer(self, stream, begin, end, hooks={}):
+        actual = 0
+        token_nb = 0
+
+        def _buffer_to_tokens(stream):
+            while True:
+                content = stream.read(CryptographyTools.READ_SIZE)
+                if len(content) == 0:
+                    break
+                yield content
+
+        f = hooks.get('buffer_to_tokens', _buffer_to_tokens)
+        # recover each sentence (encrypted Token)
+        for i in f(stream):
+            try:
+                # data decryption
+                content = self.decrypt(i)
+                if begin >= actual or begin <= actual + len(content):
+                    if actual <= begin:
+                        begin_content = begin - actual
+                    else:
+                        begin_content = 0
+                    if end is not None and end < len(content) + actual:
+                        to_yield = content[begin_content: end - actual]
+                    else:
+                        to_yield = content[begin_content:]
+                    yield to_yield
+                actual += len(content)
+                token_nb += 1
+            except CryptographyToolsException as e:
+                raise OioException('Cryptography exception: %s' % (str(e)))
+
+    def get_token_size(self, data_size):
+        remains_size = (data_size * 8) % 128
+        encrypt_entire_size = 128 - remains_size + (data_size * 8)
+        overhead_size = 128 + 8 + 64 + 256
+        return (overhead_size + encrypt_entire_size) / 8
+
+    def is_empty_key(self):
+        return False if self.crypt else True
```

Comment:
The key is not the point here. You don't want to know if there is a key, you want to know if the instance does something to the data. A name like `is_noop` would be better.